### DISCUSSION TITLE
Given a scene and a shape, how to know the bsdf(s) is specular or not?

### DISCUSSION THREAD

**Discussion by Mephisto405** [mitsuba-renderer/mitsuba3#490] [Q&A] [answered]
```
scene = mi.load_file("data/cornell-box/scene_v3.xml")
print(scene.shapes()[5].bsdf()
""" The above code returns:
TwoSided[
  brdf[0] = SmoothDiffuse[
    reflectance = SRGBReflectanceSpectrum[
      value = [[0.725, 0.71, 0.68]]
    ]
  ],
  brdf[1] = SmoothDiffuse[
    reflectance = SRGBReflectanceSpectrum[
      value = [[0.725, 0.71, 0.68]]
    ]
  ]
]
"""
```
In the above code snippet, the sixth shape in the scene has only diffuse BSDF. In this case, I want a method that returns False.
If the BSDF is specular (or other 'Dirac delta' distributions), the method should return True.
Overall, the question is about how to access the 'type' of a BSDF.

One way might be to use the __repr__() function.
But I wonder whether there is another way to do that (using some built-in Mitsuba function).

**Comment by njroussel** [comment, marked as answer]
Hi,

I think you're looking for the `BSDF.flags()` method: [BSDF.flags](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.BSDF.flags). And these are the individual flag/types supported: [mitsuba.BSDFFlags](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.BSDFFlags). You will usually want to use a combination of these with: [mitsuba.has_flag](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.has_flag).

[Here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/direct_reparam.py#L151)'s an example where we use it.

**Reply by Mephisto405** [reply]
OMG, it was so easy! Thx very much.
I didn't understand the purpose and use of flags by just looking at the documentation, but now I understand properly.
Many thanks!

**Reply by njroussel** [reply]
Yes, the documentation isn't very clear :sweat_smile:, I'll add that to our to-do list.
